core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableRecursiveUnion.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.adapter.enumerable;
+
+import org.apache.calcite.linq4j.tree.BlockBuilder;
+import org.apache.calcite.linq4j.tree.DeclarationStatement;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.linq4j.tree.Expressions;
+import org.apache.calcite.linq4j.tree.ParameterExpression;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.core.SetOp;
+import org.apache.calcite.rel.core.Union;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.schema.impl.TransientTable;
+import org.apache.calcite.util.BuiltInMethod;
+
+import java.util.List;
+
+
+/**
+ * Implementation of Recursive Union in
+ * {@link EnumerableConvention enumerable calling convention}.
+ */
+public class EnumerableRecursiveUnion extends Union implements EnumerableRel {
```

Comment:
As @vlsi said earlier I am not sure if this new operator (which is a bit particular) should extend Union since there is a risk that existing rules may match when they shouldn't. 

Apart from that I can identify two main functionalities in this operator:
1. Iterative calls to the subplan till a certain number is reached or untill there are no more results.
2. Buffer and write to a side table/collection at certain intervals (i.e., end of iteration).

I think we could try to isolate these functionalities in separate logical/physical operators with clear semantics and responsibilities which could improve encapsulation and reutilization of these operators in other use cases. 

I propose something along the following lines. For 1, we could have a Logical/EnumerableIterate operator while for 2, we could have a Logical/EnumerableSpool (maybe relevant to look at CALCITE-481) that takes care of buffering and writting to the side table. Then to satisfy the use cases so far the two operators should be combined the one on top of the other (Iterate[Spool[Subplan]]). Last thing that is missing is the seed which could be placed somwhere in the subplan and we should ensure somehow that is executed only once (using a filter?).

@rubenada If what I am proposing appears very complicated, useless, or out of context please let me know. In any case the work so far is great :)

Reply:
@zabetak Thanks for your feedback. I'll propose the following re-design to implement the "recursive union all" with re-usable modules:

Firstly, Create a new `abstract class Iterate extends BiRel` with its corresponding logical / physical operators `LogicalIterate` and `EnumerableIterate`. This Iterate will have two inputs:
- The left part will be used as `seed`, and will be evaluated once
- The right part will be used as `iteration`, and will be re-evaluated over and over until no results are produced (or until an optional `maxIter` value is reached)
This design will allow us to use the `Iterate` as main implementation module for the recursive union (which requires a non-recursive term, executed once; and a recursive term, executed over and over until no results are returned).
If, in the future, this `Iterate` needs to be used for other purposes that require just a "pure" iteration (no seed needed), it will still be possible to used it: we will just need to create an Iterate with an empty Values as left (i.e. as seed).

Secondly, to implement the data transfer from the recursive union into the "work table", a new, simple `Spool` operator will be created (see CALCITE-481). For the sake of this PR, a naive, lazy spool implementation will be enough. This operator will store in a temporary collection the values returned by its input, before propagating them. When the input returns no more values, the `Spool` will flush its temporary collection into another one (which in the case of the recursive union will be the same collection that will be used to hold the work table).

Finally, the temporary table will be a scan on table represented by a list (which will be re-initialized by the `Spool`).

To sum up, to represent a recursive union we would have an `Iterate`, with two `Spool`s inserted on top of its left (seed) and right (iteration).

To show an example, a typical recursive union like (return numbers from 1 to 10):
```
WITH RECURSIVE delta(n) AS (
    VALUES (1)
    UNION ALL
    SELECT n+1 FROM delta WHERE n < 10
)
SELECT * FROM delta
```
would be represented with the pseudo-plan:
```
Iterate
  Spool(table=delta) -- seed, executed once
    Values({1})
  Spool(table=delta) -- iteration, executed until no results are returned
    Project($0=$0+1)
      Filter($0<10)
        Scan(table=delta)
```

Reply:
I like the new design but before going forward let's see what the others have to say. Just a few additional information for those following the discussion.

There has been a small debate on the name of the RepeatUnion. `Iterate` and `RecursiveUnion` were the other names that were abandoned. 

Although `Iterate`, initially seemed to be a nice abstraction, at the end of the day the description of what it should do was a union of the results of its inputs with the particularity that we always start from the `seedRel` and we repeatedly union the `iterativeRel`.

`RecursiveUnion` also seemed nice, since it is a named also used by Postgres, and fits well with the `RECURSIVE` keyword of SQL. However, it gives the false impression that the operator does some kind of recursion which is not the case.